File: learning-algorithms/src/code/slidingwindow/LongestSubstring.java
package code.slidingwindow;

import java.util.HashMap;
import java.util.Map;

/*
 * Given a string you need to print the size of the longest possible substring that has exactly k unique characters.
 * 
 */
public class LongestSubstring {

	public static void main(String[] args) {

		System.out.println(longestSubstring("aabacbebebe", 3));
		System.out.println(longestSubstring("aaaa", 1));
	
	}

	private static int longestSubstring(String str, int k) {
		int i=0,j=0;
		int maxWindow=0;
		Map<Character, Integer> map = new HashMap<>();
		char[] charArr = str.toCharArray();
		
		while(j<charArr.length) {
			char ch = charArr[j];
			int count = map.containsKey(ch) ? map.get(ch) : 0;
			map.put(ch, count + 1 );
			if(map.size() < k) {
				j++;
			} else if(map.size() == k) {
				maxWindow = Math.max(j-i+1, maxWindow);
				j++;
			} else if(map.size() > k) {
				
				while(map.size() > k) {
					char rm = charArr[i];
					map.put(rm, map.get(rm)-1);
					if(map.get(rm) == 0) {
						map.remove(rm);
					}
					i++;
				}
				j++;
			}
			
		}
		
		return maxWindow;
	}

}

File: learning-striversheet/arrays/longestSubstring.py
def longestSubstring(str):
    maxWindow = 0
    map = {}
    i=0
    j=0

    while j<len(str):
        ch = str[j]
        count = map.get(ch, 0)
        map[ch]=count+1

        if map[ch] == 1:
            maxWindow = max(maxWindow, j-i+1) 
            j+=1
        elif map[ch] > 1:
            
            while map[ch] > 1:
                rm = str[i]
                map[rm] = map[rm]-1
                if map[rm] == 0:
                    del map[rm]
                i+=1
            j+=1            
    return maxWindow



str = 'bbbbb'
print(longestSubstring(str))   


